From 8632dd2d78d1935c3eab058017eab4f2d1702ecc Mon Sep 17 00:00:00 2001
From: Murat Sezgin <msezgin@codeaurora.org>
Date: Fri, 24 Apr 2020 10:58:42 -0700
Subject: [PATCH 228/281] inet: vlan: Added an API to return next VLAN device

The new API returns the next dev when a VLAN dev passed
in parameter i.e.
eth0.10.20 -> eth0.10 -> eth0 The new API returns eth0.10
when eth0.10.20 is passed as a parameter.

Change-Id: I743beb0672e8fd0d6562eec8ec88b1e36eec2973
Signed-off-by: Shyam Sunder <ssunde@codeaurora.org>
Signed-off-by: Murat Sezgin <msezgin@codeaurora.org>
---
 include/linux/if_vlan.h | 1 +
 net/8021q/vlan_core.c   | 7 +++++++
 2 files changed, 8 insertions(+)

diff --git a/include/linux/if_vlan.h b/include/linux/if_vlan.h
index ba73d5434684..23230663054d 100644
--- a/include/linux/if_vlan.h
+++ b/include/linux/if_vlan.h
@@ -139,6 +139,7 @@ extern void __vlan_dev_update_accel_stats(struct net_device *dev,
 				    struct rtnl_link_stats64 *stats);
 
 extern struct net_device *vlan_dev_real_dev(const struct net_device *dev);
+extern struct net_device *vlan_dev_next_dev(const struct net_device *dev);
 extern u16 vlan_dev_vlan_id(const struct net_device *dev);
 extern __be16 vlan_dev_vlan_proto(const struct net_device *dev);
 
diff --git a/net/8021q/vlan_core.c b/net/8021q/vlan_core.c
index 149c7d7a8843..9e4296aabec6 100644
--- a/net/8021q/vlan_core.c
+++ b/net/8021q/vlan_core.c
@@ -130,6 +130,13 @@ struct net_device *vlan_dev_real_dev(const struct net_device *dev)
 }
 EXPORT_SYMBOL(vlan_dev_real_dev);
 
+/* Caller is responsible to hold the reference of the returned device */
+struct net_device *vlan_dev_next_dev(const struct net_device *dev)
+{
+	return vlan_dev_priv(dev)->real_dev;
+}
+EXPORT_SYMBOL(vlan_dev_next_dev);
+
 u16 vlan_dev_vlan_id(const struct net_device *dev)
 {
 	return vlan_dev_priv(dev)->vlan_id;
-- 
2.17.1

